Bug: org.elasticsearch.test.ESIntegTestCase has randomly failing shards when using excludes in terms aggregations

In one of our `ESIntegTestCase`s, we set up an index, add a couple of documents and run a number of queries against these indices.
In most cases this works fine, with 1 exception: when using `exclude(String[] terms)` in a `TermsBuilder`, every now and then (not always!), some of the shards fail.
The shard settings of our `ESIntegTestCase` are not changed; so it sets up 1-10 shards.
Whether these aggregations aim at a field on root level or a nested document does not matter - we could reproduce the error in either case (just running the test often enough).
What seems to make a difference: It only happens when applying `exclude`; with `include` we have not seen failing shards yet.
We have not tested the regular expressions; only `String[]`.

two failure messages that we saw on the `SearchResponse` in `.getShardFailures()` (nested document / root document):
```
shard [[DT3n-RvATnO4NDMUrDo4BA][visit][1]], reason [RemoteTransportException[[node_s0][local[1]][indices:data/read/search[phase/query/id]]]; nested: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: AssertionError[startIndex=0, numBits=0]; ], cause [java.lang.AssertionError: startIndex=0, numBits=0
	at org.apache.lucene.util.LongBitSet.set(LongBitSet.java:335)
	at org.elasticsearch.search.aggregations.bucket.terms.support.IncludeExclude$TermListBackedOrdinalsFilter.acceptedGlobalOrdinals(IncludeExclude.java:179)
	at org.elasticsearch.search.aggregations.bucket.terms.GlobalOrdinalsStringTermsAggregator.getLeafCollector(GlobalOrdinalsStringTermsAggregator.java:97)
	at org.elasticsearch.search.aggregations.AggregatorBase.getLeafCollector(AggregatorBase.java:132)
	at org.elasticsearch.search.aggregations.AggregatorFactory$1$1.collect(AggregatorFactory.java:204)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectExistingBucket(BucketsAggregator.java:80)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectBucket(BucketsAggregator.java:72)
	at org.elasticsearch.search.aggregations.bucket.nested.ReverseNestedAggregator$1.collect(ReverseNestedAggregator.java:91)
	at org.elasticsearch.search.aggregations.LeafBucketCollector$3.collect(LeafBucketCollector.java:73)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectExistingBucket(BucketsAggregator.java:80)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectBucket(BucketsAggregator.java:72)
	at org.elasticsearch.search.aggregations.bucket.filter.FilterAggregator$1.collect(FilterAggregator.java:67)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectExistingBucket(BucketsAggregator.java:80)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectBucket(BucketsAggregator.java:72)
	at org.elasticsearch.search.aggregations.bucket.nested.NestedAggregator$1.collect(NestedAggregator.java:118)
	at org.elasticsearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:88)
	at org.apache.lucene.search.MultiCollector$MultiLeafCollector.collect(MultiCollector.java:174)
	at org.apache.lucene.search.AssertingLeafCollector.collect(AssertingLeafCollector.java:52)
	at org.apache.lucene.search.AssertingLeafCollector.collect(AssertingLeafCollector.java:52)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreAll(Weight.java:221)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:172)
	at org.apache.lucene.search.AssertingBulkScorer.score(AssertingBulkScorer.java:79)
	at org.apache.lucene.search.AssertingBulkScorer.score(AssertingBulkScorer.java:63)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:821)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:535)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:384)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:113)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:433)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryByIdTransportHandler.messageReceived(SearchServiceTransportAction.java:376)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryByIdTransportHandler.messageReceived(SearchServiceTransportAction.java:373)
	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

and 

```
shard [[3TcVXShdQq6NoMcYQCwbtQ][visit][5]], reason [RemoteTransportException[[node_s0][local[1]][indices:data/read/search[phase/query]]]; nested: UncheckedExecutionException[QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: AssertionError[startIndex=0, numBits=0];]; nested: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: AssertionError[startIndex=0, numBits=0]; ], cause [com.google.common.util.concurrent.UncheckedExecutionException: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: AssertionError[startIndex=0, numBits=0];
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4739)
	at org.elasticsearch.indices.cache.request.IndicesRequestCache.loadIntoContext(IndicesRequestCache.java:263)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:364)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:378)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: QueryPhaseExecutionException[Query Failed [Failed to execute main query]]; nested: AssertionError[startIndex=0, numBits=0];
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:409)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:113)
	at org.elasticsearch.indices.cache.request.IndicesRequestCache$Loader.call(IndicesRequestCache.java:302)
	at org.elasticsearch.indices.cache.request.IndicesRequestCache$Loader.call(IndicesRequestCache.java:283)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4742)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)
	... 14 more
Caused by: java.lang.AssertionError: startIndex=0, numBits=0
	at org.apache.lucene.util.LongBitSet.set(LongBitSet.java:335)
	at org.elasticsearch.search.aggregations.bucket.terms.support.IncludeExclude$TermListBackedOrdinalsFilter.acceptedGlobalOrdinals(IncludeExclude.java:179)
	at org.elasticsearch.search.aggregations.bucket.terms.GlobalOrdinalsStringTermsAggregator.getLeafCollector(GlobalOrdinalsStringTermsAggregator.java:97)
	at org.elasticsearch.search.aggregations.AggregatorBase.getLeafCollector(AggregatorBase.java:132)
	at org.elasticsearch.search.aggregations.AggregatorFactory$1$1.collect(AggregatorFactory.java:204)
	at org.elasticsearch.search.aggregations.LeafBucketCollector$3.collect(LeafBucketCollector.java:73)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectExistingBucket(BucketsAggregator.java:80)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectBucket(BucketsAggregator.java:72)
	at org.elasticsearch.search.aggregations.bucket.filter.FilterAggregator$1.collect(FilterAggregator.java:67)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectExistingBucket(BucketsAggregator.java:80)
	at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.collectBucket(BucketsAggregator.java:72)
	at org.elasticsearch.search.aggregations.bucket.nested.NestedAggregator$1.collect(NestedAggregator.java:118)
	at org.elasticsearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:88)
	at org.apache.lucene.search.MultiCollector$MultiLeafCollector.collect(MultiCollector.java:174)
	at org.apache.lucene.search.AssertingLeafCollector.collect(AssertingLeafCollector.java:52)
	at org.apache.lucene.search.AssertingLeafCollector.collect(AssertingLeafCollector.java:52)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreAll(Weight.java:221)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:172)
	at org.apache.lucene.search.AssertingBulkScorer.score(AssertingBulkScorer.java:79)
	at org.apache.lucene.search.AssertingBulkScorer.score(AssertingBulkScorer.java:63)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:821)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:535)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:384)
	... 22 more
```
The format of the response is always the same and correct; so we can still retrieve results; however, if one or two shards fail, the numbers within the response are incorrect; some documents are missing. The easiest way to check whether it worked is to check the number of failed shards.

FYI: We are using **ElasticSearch 2.3.2**

--- a/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java
@@ -92,7 +92,7 @@ public class GlobalOrdinalsStringTermsAggregator extends AbstractStringTermsAggr
         globalOrds = valuesSource.globalOrdinalsValues(ctx);
 
         if (acceptedGlobalOrdinals == null && includeExclude != null) {
-            acceptedGlobalOrdinals = includeExclude.acceptedGlobalOrdinals(globalOrds, valuesSource);
+            acceptedGlobalOrdinals = includeExclude.acceptedGlobalOrdinals(globalOrds);
         }
 
         if (acceptedGlobalOrdinals != null) {

--- a/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java
@@ -44,8 +44,6 @@ import org.elasticsearch.common.xcontent.ToXContent;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.search.DocValueFormat;
-import org.elasticsearch.search.aggregations.support.ValuesSource;
-import org.elasticsearch.search.aggregations.support.ValuesSource.Bytes.WithOrdinals;
 
 import java.io.IOException;
 import java.util.HashSet;
@@ -136,8 +134,7 @@ public class IncludeExclude implements Writeable, ToXContent {
     }
 
     public abstract static class OrdinalsFilter {
-        public abstract LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals, ValuesSource.Bytes.WithOrdinals valueSource)
-                throws IOException;
+        public abstract LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals) throws IOException;
 
     }
 
@@ -154,7 +151,7 @@ public class IncludeExclude implements Writeable, ToXContent {
          *
          */
         @Override
-        public LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals, ValuesSource.Bytes.WithOrdinals valueSource)
+        public LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals)
                 throws IOException {
             LongBitSet acceptedGlobalOrdinals = new LongBitSet(globalOrdinals.getValueCount());
             TermsEnum globalTermsEnum;
@@ -180,7 +177,7 @@ public class IncludeExclude implements Writeable, ToXContent {
         }
 
         @Override
-        public LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals, WithOrdinals valueSource) throws IOException {
+        public LongBitSet acceptedGlobalOrdinals(RandomAccessOrds globalOrdinals) throws IOException {
             LongBitSet acceptedGlobalOrdinals = new LongBitSet(globalOrdinals.getValueCount());
             if (includeValues != null) {
                 for (BytesRef term : includeValues) {
@@ -189,7 +186,7 @@ public class IncludeExclude implements Writeable, ToXContent {
                         acceptedGlobalOrdinals.set(ord);
                     }
                 }
-            } else {
+            } else if (acceptedGlobalOrdinals.length() > 0) {
                 // default to all terms being acceptable
                 acceptedGlobalOrdinals.set(0, acceptedGlobalOrdinals.length());
             }

--- a/core/src/test/java/org/elasticsearch/search/aggregations/support/IncludeExcludeTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/support/IncludeExcludeTests.java
@@ -1,0 +1,128 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.RandomAccessOrds;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.LongBitSet;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.bucket.terms.support.IncludeExclude;
+import org.elasticsearch.search.aggregations.bucket.terms.support.IncludeExclude.OrdinalsFilter;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.TreeSet;
+
+public class IncludeExcludeTests extends ESTestCase {
+
+    public void testEmptyTermsWithOrds() throws IOException {
+        IncludeExclude inexcl = new IncludeExclude(
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))),
+                null);
+        OrdinalsFilter filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        LongBitSet acceptedOrds = filter.acceptedGlobalOrdinals(DocValues.emptySortedSet());
+        assertEquals(0, acceptedOrds.length());
+
+        inexcl = new IncludeExclude(
+                null,
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))));
+        filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        acceptedOrds = filter.acceptedGlobalOrdinals(DocValues.emptySortedSet());
+        assertEquals(0, acceptedOrds.length());
+    }
+
+    public void testSingleTermWithOrds() throws IOException {
+        RandomAccessOrds ords = new RandomAccessOrds() {
+
+            boolean consumed = true;
+
+            @Override
+            public void setDocument(int docID) {
+                consumed = false;
+            }
+
+            @Override
+            public long nextOrd() {
+                if (consumed) {
+                    return SortedSetDocValues.NO_MORE_ORDS;
+                } else {
+                    consumed = true;
+                    return 0;
+                }
+            }
+
+            @Override
+            public BytesRef lookupOrd(long ord) {
+                assertEquals(0, ord);
+                return new BytesRef("foo");
+            }
+
+            @Override
+            public long getValueCount() {
+                return 1;
+            }
+
+            @Override
+            public long ordAt(int index) {
+                return 0;
+            }
+
+            @Override
+            public int cardinality() {
+                return 1;
+            }
+        };
+        IncludeExclude inexcl = new IncludeExclude(
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))),
+                null);
+        OrdinalsFilter filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        LongBitSet acceptedOrds = filter.acceptedGlobalOrdinals(ords);
+        assertEquals(1, acceptedOrds.length());
+        assertTrue(acceptedOrds.get(0));
+
+        inexcl = new IncludeExclude(
+                new TreeSet<>(Collections.singleton(new BytesRef("bar"))),
+                null);
+        filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        acceptedOrds = filter.acceptedGlobalOrdinals(ords);
+        assertEquals(1, acceptedOrds.length());
+        assertFalse(acceptedOrds.get(0));
+
+        inexcl = new IncludeExclude(
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))),
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))));
+        filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        acceptedOrds = filter.acceptedGlobalOrdinals(ords);
+        assertEquals(1, acceptedOrds.length());
+        assertFalse(acceptedOrds.get(0));
+
+        inexcl = new IncludeExclude(
+                null, // means everything included
+                new TreeSet<>(Collections.singleton(new BytesRef("foo"))));
+        filter = inexcl.convertToOrdinalsFilter(DocValueFormat.RAW);
+        acceptedOrds = filter.acceptedGlobalOrdinals(ords);
+        assertEquals(1, acceptedOrds.length());
+        assertFalse(acceptedOrds.get(0));
+    }
+
+}

